### PR TITLE
[Flax] Add Beam Search

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This adds beam search for Flax. Aggressive integration tests for Bart-large-cnn is added.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Circle CI error seem unrelated: ```OSError: /home/circleci/.local/lib/python3.7/site-packages/torch_scatter/_scatter_cpu.so: undefined symbol: _ZNK2at6Tensor6deviceEv```

**Comment by patrickvonplaten** [comment]
Rebase from https://github.com/huggingface/transformers/pull/12181

File: tests/test_modeling_flax_bart.py [hunk 1]
```diff
@@ -411,7 +412,66 @@ def decode_jitted(decoder_input_ids, decoder_attention_mask, encoder_outputs):
     def test_model_from_pretrained(self):
         for model_class_name in self.all_model_classes:
             model = model_class_name.from_pretrained("facebook/bart-base", from_pt=True)
-            # FlaxBartForSequenceClassification expects eos token in input_ids
+            # FlaxBartForSequenceClassification expects eos tokenizeren in input_ids
```
Comment (sgugger) [inline tests/test_modeling_flax_bart.py:415]:
```suggestion
            # FlaxBartForSequenceClassification expects eos token in input_ids
```
Weird change!

**Review by sgugger — APPROVED** [review]
Impressive! Thanks a lot for adding this!

File: src/transformers/generation_flax_utils.py [hunk 2]
```diff
@@ -43,9 +48,8 @@ class FlaxGreedySearchOutput(ModelOutput):
 
 
     Args:
-        sequences (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`):
-            The generated sequences. If all batches finished early due to the :obj:`eos_token_id`, :obj:`sequences` is
-            padded to :obj:`max_length`.
+        sequences (:obj:`torch.LongTensor` of shape :obj:`(batch_size, max_length)`):
```
Comment (patil-suraj) [inline src/transformers/generation_flax_utils.py:51]:
```suggestion
        sequences (:obj:`jax_xla.DeviceArray` of shape :obj:`(batch_size, max_length)`):
```

File: src/transformers/generation_flax_utils.py [hunk 3]
```diff
@@ -59,18 +63,34 @@ class FlaxSampleOutput(ModelOutput):
 
     Args:
         sequences (:obj:`torch.LongTensor` of shape :obj:`(batch_size, max_length)`):
```
Comment (patil-suraj) [inline src/transformers/generation_flax_utils.py:65]:
```suggestion
        sequences (:obj:`jax_xla.DeviceArray` of shape :obj:`(batch_size, max_length)`):
```

File: src/transformers/generation_flax_utils.py [hunk 4]
```diff
@@ -59,18 +63,34 @@ class FlaxSampleOutput(ModelOutput):
...
+class FlaxBeamSearchOutput(ModelOutput):
+    """
+    Flax Base class for outputs of decoder-only generation models using greedy search.
+
+
+    Args:
+        sequences (:obj:`torch.LongTensor` of shape :obj:`(batch_size, max_length)`):
+            The generated sequences.
+        scores (:obj:`torch.LongTensor` of shape :obj:`(batch_size,)`):
+            The scores (log probabilites) of the generated sequences.
```
Comment (patil-suraj) [inline src/transformers/generation_flax_utils.py:82]:
```suggestion
        sequences (:obj:`jax_xla.DeviceArray` of shape :obj:`(batch_size, max_length)`):
            The generated sequences.
        scores (:obj:`jax_xla.DeviceArray` of shape :obj:`(batch_size,)`):
            The scores (log probabilites) of the generated sequences.
```

File: tests/test_modeling_flax_bart.py [hunk 5]
```diff
@@ -415,3 +416,62 @@ def test_model_from_pretrained(self):
...
+
+        dct = tokenizer.batch_encode_plus(
```
Comment (patil-suraj) [inline tests/test_modeling_flax_bart.py:450]:
```suggestion
        dct = tokenizer(
```

**Review by patil-suraj — APPROVED** [review]
Really amazing!

**Pull request merged**